### PR TITLE
Random fixes and tweaks

### PR DESCRIPTION
**Pull request by alexdewar**
This is just a smattering of random fixes and small changes cherry-picked from other branches I'm working on. I thought it might be easier just to merge these little bits all at once.

Bigger changes:
* Swap templated get{Pose,Position,Attitude}() functions for non-templated versions -- it's cleaner and we can cast between e.g. Pose objects anyway
* For snapshot_bot, let the user drive it with the joystick when it's in idle state, so we don't have to pick it up and carry it back to the start position :-)

Other than that it should be pretty self-explanatory and non-controversial.

### CODE REVIEW

File: examples/ant_agent/ant_agent.cc [hunk 1]
```diff
@@ -62,7 +62,7 @@ bob_main(int, char **argv)
         // Poll joystick
         joystick.update();
 
-        const auto pose = agent.getPose<meter_t, degree_t>();
+        const decltype(lastPose) pose = agent.getPose();
```
[the remark this answers is not in the document]
Comment (alexdewar) [reply examples/ant_agent/ant_agent.cc]: It just makes ``pose`` the same type as ``lastPose``. Can't remember why I changed it though :-/

File: include/common/has_pose.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#pragma once
+
+namespace BoBRobotics {
+template<class Derived>
+class HasPose {
+public:
+    auto getAttitude() {
+        auto derived = reinterpret_cast<Derived *>(this);
```
[the remark this answers is not in the document]
Comment (alexdewar) [reply include/common/has_pose.h]: Ah yeah. It looks like you're right.

File: include/vicon/udp.h [hunk 3]
```diff
@@ -349,15 +321,15 @@ class UDPClient
...
+    size_t getNumObjects() const
     {
         waitUntilConnected();
         std::lock_guard<std::mutex> guard(m_ObjectMutex);
         return m_ObjectData.size();
     }
 
     //! Get current pose information for specified object
-    ObjectDataType getObjectData(const std::string &name)
+    const ObjectDataType &getObjectData(const std::string &name) const
```
[the remark this answers is not in the document]
Comment (neworderofjamie) [reply include/vicon/udp.h]: maybe add a comment making it clear why it's returned by value - I seem to remember this getting changed before

File: include/common/has_pose.h [hunk 4]
```diff
@@ -0,0 +1,17 @@
+#pragma once
+
+namespace BoBRobotics {
+template<class Derived>
+class HasPose {
```
[the remark this answers is not in the document]
Comment (neworderofjamie) [reply include/common/has_pose.h]: as we're in the advantageous position where there's not a lot of code, outside of this repo, using this I think breaking code for the sake of not adding this layer of complexity is worth it but up to you

File: examples/ant_agent/ant_agent.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (neworderofjamie) [reply examples/ant_agent/ant_agent.cc]: As long as ``Pose3`` doesn't have a weird == overload, the compiler will complain if somehow ``agent.getPose()`` returns a different type

[the remark this answers is not in the document]
Comment (alexdewar) [reply examples/ant_agent/ant_agent.cc]: You can implicitly cast between ``Pose`` types now so it should work in any case

File: include/common/has_pose.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (alexdewar) [reply include/common/has_pose.h]: True. Let's drop it so we don't add unnecessary boilerplate

File: include/vicon/udp.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexdewar) [reply include/vicon/udp.h]: :+1:

**Pull request merged**